### ISSUE TITLE
CC3 Engine Code double size from CC2

### ISSUE THREAD

**Issue by Ronsku** [cocos/cocos-engine#10354]
In CC3 the cc.js engine file seems to be quite a bit larger than in CC2. It's still huge after disabling all the features I don't need.
Are you looking into making it smaller or isn't it possible?

CC3 (2200 KB)
![image](https://user-images.githubusercontent.com/7983358/158453059-1bae09c3-e25d-4639-8c2e-9e6822da14bf.png)

CC2 (1011 KB, Over 100% smaller!)
![image](https://user-images.githubusercontent.com/7983358/158453083-1f7e9eae-57bd-41b0-a0ec-379f6fd42132.png)

And this one from CC3 is with almost every feature turned off:
![image](https://user-images.githubusercontent.com/7983358/158453101-802b0c69-4910-4ab3-b633-fd8060d51a3e.png)

I'm guessing there are something from 3D that is not cleaned up in the engine, even though you create a fully 2D game? This has a big effect on total bundle size of games.

**Comment by minggo** [comment]
@Ronsku yep, we are still trying do more effort on it. Can i know where will you publish your games?

**Comment by Ronsku** [comment]
Thanks @minggo, great to hear!
I only publish for **web-mobile** platform, let me know if I can help testing in any way.

**Comment by minggo** [comment]
Thanks @Ronsku , i will ask your help if needed.
Can i know more about the web-mobile platform? What's the limitation of the platform, for example, size limitation. And how can you earn money from it?

**Comment by Ronsku** [comment]
@minggo, No limitations really, just like any website or software you want it to be loading as fast as possible to your users. Doubling the size of the engine causes bigger loading times and a worse user-experience. Earnings comes from user satisfaction and from the same logic as loading a webpage. If the webpage you are going to is loading for too long you might leave it and the company would lose a potential customer.

I think loading times are always very important for a good UX. You want to play, not watch a loading screen.

**Comment by minggo** [comment]
@Ronsku thanks for the information. How long it takes based on your testing? And what's the loading time is acceptable?

**Comment by Ronsku** [comment]
> @Ronsku thanks for the information. How long it takes based on your testing? And what's the loading time is acceptable?

That's a very difficult question to answer, since it is all based on server and user connection and also user's device that needs to run the code, launch the Web Canvas etc.
The smaller the better without of course sacrificing features. This is why the "Feature Cropping" is a thing in Cocos Creator, to make the final bundle as small as possible while maintaining the features you've selected for your game. This way you wouldn't for example need to download all 3D features in a 2D game or DragonBones, unless you use it. Just as examples.

Problem in this case is that the CC3 final built code with most features cropped out is still double the side of the final build of a CC2 game.

Thank you @pandamicro for picking this up. Highly appreciated!

**Comment by Ronsku** [comment]
@pandamicro and @minggo 
I have confirmed that CC package size has gone up again in CC 3.5 slightly
![image](https://user-images.githubusercontent.com/7983358/167851945-91ea34b5-51f9-477d-b71f-934980e637fd.png)
from
CC2: **1011Kb**
CC 3.4: **2,200kb**
CC 3.5: **2,247kb**
This is also with removing Marionette system from previous test
![image](https://user-images.githubusercontent.com/7983358/167852188-87d586b1-a3f6-4f97-8b28-35506acd43d4.png)
Almost everything removed and package still big.

I'm not sure what is making it so massive, but it hurts loading times a lot.

**Comment by SantyWang** [comment]
Yes, we have identified some optimization points, such as eliminating built-in effects and optimizing the decorators, and in version 3.6, we will optimize these part

**Comment by Ronsku** [comment]
> Yes, we have identified some optimization points, such as eliminating built-in effects and optimizing the decorators, and in version 3.6, we will optimize these part

Amazing, can't wait! I will be a very big boost to the game loading :)

**Comment by smitpatel88** [comment]
@holycanvas Does this help in native env as well?

**Comment by Ronsku** [comment]
@holycanvas I really hope this ticket will be taken seriously, because it is actually causing a lot of problems for me currently.

Having over 2x size on cc.js doesn't only cause 2x loading downloading time, but the worst part is that especially on slower devices parsing of huge JavaScript files is really costly.

I have huge difference between loading CC2 and CC3 games. I was even struggling a bit in CC2 to get good loading times. Now in CC3 it just takes way too long with this massive 2.2Mb chunk of JavaScript.
Of course I run it through GZIP, but it's still double the size from CC2.

Can you help me understand what is the main reason behind the size?
Are there any way to manually clean up features not used?

**Comment by Ronsku** [comment]
> Yes, we have identified some optimization points, such as eliminating built-in effects and optimizing the decorators, and in version 3.6, we will optimize these part

How about "Feature Cropping"? It doesn't seem to affect the bundle size if you check out most of the packages form the feature cropping in project settings.

For example in 2D games, I'm just guessing that the whole 3D library is packaged with it, even though it is not selected in the feature cropping.

**Comment by Ronsku** [comment]
Some Google Lighthouse stats, on a game with most features removed 😬 

Non-GZIP
![image](https://user-images.githubusercontent.com/7983358/171967496-17fc637b-ad87-4fee-b9d7-a1d0defe22b5.png)
![image](https://user-images.githubusercontent.com/7983358/171967438-d3d1c48a-d679-45f8-9331-ba7ceaf72513.png)
![image](https://user-images.githubusercontent.com/7983358/171967461-392d06c6-1eeb-4be4-95d3-99f3127f8532.png)

GZIP:ed
![image](https://user-images.githubusercontent.com/7983358/171967882-f08494b4-91ff-40aa-ae05-678939927b26.png)
![image](https://user-images.githubusercontent.com/7983358/171967837-6543a2d9-5a93-4b1a-957c-3d5b21e45174.png)
![image](https://user-images.githubusercontent.com/7983358/171967941-fd287783-06f5-44c8-88d7-efa6eaf707d1.png)

**Comment by Ronsku** [comment]
Same on **CC 2** Gzip:ed, not great, but much better:
![image](https://user-images.githubusercontent.com/7983358/171968268-df7267d6-ebdf-494e-bc2b-ccf61ad2fb9c.png)
![image](https://user-images.githubusercontent.com/7983358/171968213-3d8ae760-4382-45ff-9279-ca6517c10169.png)
![image](https://user-images.githubusercontent.com/7983358/171968329-9b16a927-b505-4360-a793-94a5f41f8883.png)

**Comment by shrinktofit** [comment]
@Ronsku  Have you tried the build setting "Erase module structure(experimental)"? That option will flattern import chain.

**Comment by Ronsku** [comment]
@shrinktofit No I had not tried that one before. I tried it now with an empty project. Here are the results:
Feature Cropping:
![image](https://user-images.githubusercontent.com/7983358/172342013-9469bb0b-057d-4229-b3e7-d7810068f8a9.png)

Build Settings WITHOUT "Erase module structure(experimental)"
![image](https://user-images.githubusercontent.com/7983358/172341930-1de8d710-902e-44bc-b827-a8d309bf992b.png)
![image](https://user-images.githubusercontent.com/7983358/172342075-e2d7d2ad-a147-4a98-8a00-c651ae44ef12.png)

Build Settings WITH "Erase module structure(experimental)"
![image](https://user-images.githubusercontent.com/7983358/172342410-3dbdb8e0-d9e2-4650-9488-002bb715028e.png)
![image](https://user-images.githubusercontent.com/7983358/172342426-d6d5eeb4-32d0-4331-b51c-5ab2370db02f.png)

Size of the Engine code is exactly the same 😞

**Comment by shrinktofit** [comment]
@Ronsku Well this option is dedicated to project scripts, not engine :)

**Comment by Ronsku** [comment]
@shrinktofit I see, but didn't help with loading speed for me at least. Engine size is the issue

**Comment by pandamicro** [comment]
In v3.6, we will remove effects.js from the engine to achieve two goals
1. Reduce engine package size by 20% - 30%
2. Improve engine initialize speed by only loading necessary shaders to GPU

This is also described in https://github.com/cocos/cocos-engine/issues/10529, so we will complete the last in v3.6, and I'm moving this issue to v3.7 so that we can keep tracking further optimizations

@holycanvas How is the progression of this task ?

**Comment by Ronsku** [comment]
> @holycanvas How is the progression of this task ?

@holycanvas Is this ready for 3.6 already. Am I able to see fetch it as a custom engine already or is this a build in editor feature somehow?

**Comment by SantyWang** [comment]
> @holycanvas Is this ready for 3.6 already

@Ronsku Yes, In 3.6, all builtin effects had been removed from engine

**Comment by Ronsku** [comment]
in 3.6.0 beta we've gone down from 2.2mb to 1.7mb in size of the built engine (77.3% of previous).
When GZIP:ed it went down from 743Kb to 603Kb (81.2% of previous)

Compared to CC2 the difference is still big:
CC2: Gzip:ed: 380Kb
CC3: Gzip:ed: 603Kb (59% bigger)

**Comment by SantyWang** [comment]
Thanks for your testing. In v3.6, we have optimized some effect. and In v3.7, We will optimize the module and decorator and so on.

**Comment by Ronsku** [comment]
Update from CC 3.7.0
I just ran a very basic 2D game build with no 3D features in the feature cropping and only a few 2D things picked.
The engine size is even bigger than before 2330 KB compared to CC2 which currently run around 990 KB

CC3:
![image](https://user-images.githubusercontent.com/7983358/216341660-ace2f462-719c-47d0-add2-1138a342cbd6.png)

CC2:
![image](https://user-images.githubusercontent.com/7983358/216341714-7e44c4f3-7d4a-4dcf-bbf8-ad7e841faceb.png)

**Comment by Ronsku** [comment]
Update on CC 3.7.0

By Removing
- 2D Physics System
- 2D Intersection Algorithms
- 2D Particle System

We come down to 1796 KB, but that's still Double the size of CC2 and loading is not only download size, but also weaker devices Evaluation and CPU time going through all of that code.

![image](https://user-images.githubusercontent.com/7983358/216342856-cf3f3edb-1bae-4d44-b303-306072ab5ca4.png)
![image](https://user-images.githubusercontent.com/7983358/216342882-b3b6d065-9945-4873-8e78-25f36ea34380.png)

**Comment by Ronsku** [comment]
@PPpro @holycanvas @pandamicro
"[PPpro](https://github.com/PPpro) removed this from the [V3.7](https://github.com/cocos/cocos-engine/milestone/5) milestone [on Dec 5, 2022](https://github.com/cocos/cocos-engine/issues/10354#event-7950104709)"

Are there any plans to bring this back?
This and scene swapping performance https://github.com/cocos/cocos-engine/issues/11503, are the main reasons I have to stick with CC2 instead of CC3.

I need to be able to create 2D games that loads fast and efficiently. CC2 is already much slower and bigger than than for example PixiJS, but still in an acceptable range. While CC3 becomes so slow and heavy that it's not an option anymore :(.

I and a lot of other people really wants to swap to CC3 to get all the other benefits from it, but we're unable due to this.

**Comment by minggo** [comment]
@Ronsku we are improving modularization these days. Then developers can use features they want. It will be released in v3.8, which is planned to be released in June.

**Comment by Ronsku** [comment]
3.7.2 Update, engine size has remained close to 2mb (2x size of CC2)
![image](https://user-images.githubusercontent.com/7983358/227478063-63758bf3-8b02-438f-83e6-ab110f669705.png)
Result of build:
![image](https://user-images.githubusercontent.com/7983358/227478183-578354a4-35b2-4306-8850-b9e180c4d652.png)

**Comment by PPpro** [comment]
Sorry, we adjusted this issue plan to milestrone 3.9

**Comment by Ronsku** [comment]
> Sorry, we adjusted this issue plan to milestrone 3.9

Okay, thanks for the update!
Hoping so much for a great size reduction and speed of initial load, so that I can swap from CC2 to CC3 before it becomes CC4.

Thanks, let me know if I can help in any way!

**Comment by hubluesky** [comment]
麻烦一起看看3.x各个版本之后的包体增加，应该是相同的问题。
https://forum.cocos.org/t/topic/145082

**Comment by Ronsku** [comment]
> 麻烦一起看看3.x各个版本之后的包体增加，应该是相同的问题。 https://forum.cocos.org/t/topic/145082

Thanks for posting this on the forums as well! Any attention to this will help😊.
Engine size and initial load times are so important. I understand that 3D games requires a bigger engine bundle, but there are a lot of people that still want to make 2D games. I first liked the idea of bringing Cocos 3D into Cocos Creator, which back in CC2 was only for 2D games. But I could have never guessed that we would ruin 2D game creation by having bundle size and load times 2x as long for the same thing.

I'm just hoping that we could get the latest Cocos Creator versions bundle size and loading times to be as good or better as it was to create 2D games in CC2 many years back.

I'm still forced to use 2.4.11.

**Comment by jareguo** [comment]
Re: https://forum.cocos.org/t/topic/149894

**Comment by Ronsku** [comment]
Update from CC **3.8.0**, vs latest CC2 version **2.4.11**.

Very Stripped 2D Build in 3.8.0:
![image](https://github.com/cocos/cocos-engine/assets/7983358/555e5cb4-770d-4d43-992f-ce90946c7a75)

I think things were moved out from the bundle, but the bundle is still relatively the same size or slightly bigger?
Compared to version 3.7.2:
**3.7.2:** CC bundle is **1,806 KB**
**3.8.0:** CC bundle is **1,825 KB** while it looks like Spine has even been stripped out and adds an additional 701 KB?!
![image](https://github.com/cocos/cocos-engine/assets/7983358/104433c3-42c1-4d1d-aa7b-8a17f5274b37)

**2.4.11** while CC2 stays at **991 KB**

Here are the statistics of just the CC bundle between CC2 and CC3, without taking into consideration the 701 KB Spine file.

![image](https://github.com/cocos/cocos-engine/assets/7983358/f99424b3-82e4-48a9-b570-6ae8a181636d)

I'm getting really worried getting stuck in legacy CC2 forever or having to look for alternatives, which I really wouldn't like to do 😨?!

**Comment by Ronsku** [comment]
> Sorry, we adjusted this issue plan to milestrone 3.9

Praying for 3.9 with everything I have! 🙏

**Comment by heyanfei** [comment]
在3.8.1-0913版本上验证，包体确实如开发者反馈的那样偏大，且spine已移除出去了（740KB），包体还是没有变化，还是1.8M，这个额外的增加是？

**Comment by jareguo** [comment]
We will continue to optimize the engine's packages in 3.8.x. But it's really hard to fully optimize to 2.x level

**Comment by PPpro** [comment]
> 3.7.2: CC bundle is 1,806 KB
3.8.0: CC bundle is 1,825 KB while it looks like Spine has even been stripped out and adds an additional 701 KB?!

the spine asm.js module is a fallback for spine wasm module, which is another part of the added package size.  
we've provided an build option to cull the asm.js module to reduce package size, please have a try @Ronsku 
![image](https://github.com/cocos/cocos-engine/assets/17872773/6b1b8f03-342f-4422-a1bd-aae3d2103bf9)

**Comment by Ronsku** [comment]
@jareguo Thank you so much, I believe its a big challenge, but 1.9Mb engine size + features like spine is so big for smaller games, where you would like to ensure a fast loading.

Compared to the size of for example pixi.js v6.4.2 is 479.1 kB (minified), and 125.6 kB when compressed using GZIP. The latest versions even slightly smaller.

That makes CC2.4.8 about **2.8x bigger** and CC3.8.0 **5x bigger** than pixi.js using GZIP.
I understand they are very different, but it just feels so unfortunate that the difference is so massive, when I otherwise love Cocos Creator 😞.

> the spine asm.js module is a fallback for spine wasm module, which is another part of the added package size.

Thanks for your message @PPpro, here are the results in CC 3.8.0:

Without checked "Cull Engine Asm.js Module":
- Both .wasm and .js are loading on web assembly supported browser.
- Adding 1124Kb uncompressed size
![image](https://github.com/cocos/cocos-engine/assets/7983358/4f44e726-5621-4b08-8897-7109239f8fb3)

With checked "Cull Engine Asm.js Module":
- Only .wasm loading
- Adding 406Kb uncompressed size.
![image](https://github.com/cocos/cocos-engine/assets/7983358/3219d57c-c719-4938-8f8f-a2788634f1fc)

Engine itself in both cases adding an additional whopping 1.9Mb uncompressed size. Massive for 2D use 😓.

**Comment by Ronsku** [comment]
Just to update this issue. I ran similar tests on 3.8.1 and 3.8.2 as my above 3.8.0 output.

**Comment by minggo** [comment]
The goal for 3.8.4 is to make the package size less or equal 3.8.0, and we will continue reducing package size in all versions.

**Comment by Ronsku** [comment]
I read about so many "Code size"-optimizations in CC 3.8.6, so I got hyped. But we're back to same 3.8.0 size as it seems.

<img width="649" height="212" alt="Image" src="https://github.com/user-attachments/assets/8d217562-acbf-44bf-9174-4089c5b6b8d8" />

Web-mobile empty project build with most features toggled off is 2.5mb load gzipped. This is as big as a ready game in CC2 including a lot of game assets.

<img width="993" height="230" alt="Image" src="https://github.com/user-attachments/assets/17b9b506-363c-4839-abbf-349bc59b5404" />

If WeChat Mini Games can get a package size down to 160KB, how can that be done for Web-Mobile games @minggo?
